### ISSUE TITLE
severity(warning) should probably not use an `Error:` introducer

### ISSUE THREAD

**Issue by CAD97** [zkat/miette#139]
Example:

```rust
use miette::{Diagnostic, NamedSource, Result, SourceSpan};
use thiserror::Error;

#[derive(Error, Debug, Diagnostic)]
#[error("oops!")]
#[diagnostic(
    code(oops::my::bad),
    url(docsrs),
    help("try doing it better next time?"),
    severity(warning)
)]
struct MyBad {
    #[source_code]
    src: NamedSource,
    #[label("This bit here")]
    bad_bit: SourceSpan,
}

fn main() -> Result<()> {
    let src = "source\n  text\n    here".to_string();

    Err(MyBad {
        src: NamedSource::new("bad_file.rs", src),
        bad_bit: (9, 4).into(),
    })?;

    Ok(())
}
```

```
Error: oops::my::bad (link)

  ⚠ oops!
   ╭─[bad_file.rs:1:1]
 1 │ source
 2 │   text
   ·   ──┬─
   ·     ╰── This bit here
 3 │     here
   ╰────
  help: try doing it better next time?
```

![image](https://user-images.githubusercontent.com/5992217/159103152-90b4f475-7ff0-4fe7-ac2f-74e9199b2858.png)

Similarly with severity(advice).

A more realistic example with `#[related]`:

<details>

```rust
use miette::{Diagnostic, Result, SourceSpan};
use std::{error::Error, fmt::Display};
use thiserror::Error;

#[derive(Error, Debug, Diagnostic)]
#[diagnostic(code(D0000), severity(error))]
#[error("I won't allow this")]
struct SourceError {
    #[label]
    span: SourceSpan,
}

#[derive(Error, Debug, Diagnostic)]
#[diagnostic(code(D4000), severity(warning))]
#[error("I don't like this")]
struct SourceWarning {
    #[label]
    span: SourceSpan,
}

#[derive(Error, Debug, Diagnostic)]
#[diagnostic(code(D8000), severity(advice))]
#[error("This could be better")]
struct SourceAdvice {
    #[label]
    span: SourceSpan,
}

#[derive(Debug)]
struct DynDiagnostic {
    inner: Box<dyn Diagnostic + Send + Sync + 'static>,
}

impl Display for DynDiagnostic {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.inner.fmt(f)
    }
}

impl Error for DynDiagnostic {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        self.inner.source()
    }
}

impl Diagnostic for DynDiagnostic {
    fn code<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.inner.code()
    }

    fn severity(&self) -> Option<miette::Severity> {
        self.inner.severity()
    }

    fn help<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.inner.help()
    }

    fn url<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.inner.url()
    }

    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
        self.inner.source_code()
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
        self.inner.labels()
    }

    fn related<'a>(&'a self) -> Option<Box<dyn Iterator<Item = &'a dyn Diagnostic> + 'a>> {
        self.inner.related()
    }
}

impl DynDiagnostic {
    fn new(diagnostic: impl Diagnostic + Send + Sync + 'static) -> Self {
        DynDiagnostic {
            inner: diagnostic.into(),
        }
    }
}

#[derive(Error, Debug, Diagnostic)]
#[diagnostic(severity(error))]
#[error("compilation failed")]
struct SourceDiagnostics {
    #[source_code]
    src: String,
    #[related]
    diagnostics: Vec<DynDiagnostic>,
}

fn main() -> Result<()> {
    let src = "\
this is an error
this is a warning
this is an advice
";

    let error = SourceError {
        span: (11..16).into(),
    };
    let warning = SourceWarning {
        span: (27..34).into(),
    };
    let advice = SourceAdvice {
        span: (46..52).into(),
    };

    let diagnostics = SourceDiagnostics {
        src: src.into(),
        diagnostics: vec![
            DynDiagnostic::new(error),
            DynDiagnostic::new(warning),
            DynDiagnostic::new(advice),
        ],
    };

    Err(diagnostics)?;
    Ok(())
}
```

```
Error:
  × compilation failed

Error: D0000

  × I won't allow this
   ╭─[1:1]
 1 │ this is an error
   ·            ─────
 2 │ this is a warning
   ╰────
Error: D4000

  ⚠ I don't like this
   ╭─[1:1]
 1 │ this is an error
 2 │ this is a warning
   ·           ───────
 3 │ this is an advice
   ╰────
Error: D8000

  ☞ This could be better
   ╭─[2:1]
 2 │ this is a warning
 3 │ this is an advice
   ·            ──────
   ╰────
```
</details>

![image](https://user-images.githubusercontent.com/5992217/159103783-fe9d9bad-bcc1-4cb4-9809-8553cddaeaac.png)

**Comment by zkat** [comment]
This is, unfortunately, hard coded into Rust's main() -> Result handling. You have to manually report errors instead of using the global hook if you want to get rid of this :(

**Comment by CAD97** [comment]
What about the `#[related]` case, though? The `Error:` introducer there is added by the hook, is it not?

Obviously, an error returned from main is an error. I'm just using it as a simple way to print the diagnostic.

**Comment by CAD97** [comment]
https://github.com/zkat/miette/blob/main/src/handlers/graphical.rs#L261

**Comment by zkat** [comment]
Oh I forgot about this. Yeah, that one I just did for "consistency". I think it would be weird for a diagnostic to be preceded with "Error:" when it's the top one, and "Warning:" when it happens to be a `related` one?

**Comment by CAD97** [comment]
I suppose I can always make my own `Diagnostics` container which collates a set of diagnostics in a different manner.

Currently what I'm using in `tracing-filter` is `fn Filter::parse(spec; impl AsRef<str> + Into<String>) -> (Option<Self>, Option<ErrReport>);` The `ErrReport` is then either a [single error](https://github.com/CAD97/tracing-filter/blob/a9cbb260aa38d3eb6a6c23c2f193f8c1863ffe71/src/simple/mod.rs#L223-L238)[^1] or a [set of warnings](https://github.com/CAD97/tracing-filter/blob/a9cbb260aa38d3eb6a6c23c2f193f8c1863ffe71/src/simple/mod.rs#L171-L177). It's then up to the library user how they want to display [the resulting error/warnings](https://github.com/CAD97/tracing-filter/blob/main/tests/snapshots/lib__snapshots__report%40target.env_filter.snap)[^2].

I honestly don't know what the best default solution would be here; it's also decently likely that any serious compiler would want to have use their own custom diagnostic collection type for more control anyway. What I think would make the picture clearer would be if/when miette gets support for emitting structured suggestions/fixes, and considering other use cases that want a proper tree of `Diagnostic` rather than a flat list.

[^1]: This format *exactly* matches env_logger's, so can't report multiple errors nor an error and warnings. The other format will likely be able to emit both multiple errors and warnings at the same time, but it will always be contextually clear from control flow whether the root `ErrReport` is an actual error (parsing/compiling failed) or a warning (parsing/compiling limped along).

[^2]: I think the recommendation is going to be to `error!` the `ErrReport`[^3] after the log sink is installed; the compilation continuing with warnings is desirable for "best effort" progress, but any warnings in the provided log filter is an indication of user error.

[^3]: In the future I might end up stripping the `ErrReport` container and providing `Vec<dyn Diagnostic>` instead, so the diagnostics can be structurally reported as a [`Listable`](https://docs.rs/valuable/latest/valuable/trait.Listable.html). In the (far) future it might be fun to play with being able to use e.g. the JSON style renderer to trace a [structured valuable diagnostic](https://docs.rs/valuable/latest/valuable/derive.Valuable.html)[^4] to "more structured" tracing sinks, but use the graphical rendering for "more user facing" tracing sinks. Some interesting thoughts.

[^4]: Clarification[^5] note: valuable support in tracing is still experimental and unstable. However I have faith that some form of it will come eventually!

[^5]: Wow I've overused footnotes here.

**Comment by CAD97** [comment]
I'm also thinking about things like this rustc warning:

```
warning: unreachable `pub` item
   --> tracing-error\src\lib.rs:228:21
    |
228 |     pub use crate::{ExtractSpanTrace as _, InstrumentError as _, InstrumentResult as _};
    |     ---             ^^^^^^^^^^^^^^^^^^^^^
    |     |
    |     help: consider restricting its visibility: `pub(crate)`
    |
note: the lint level is defined here
   --> tracing-error\src\lib.rs:191:5
    |
191 |     unreachable_pub,
    |     ^^^^^^^^^^^^^^^
    = help: or consider exporting it for use by other crates
```

The two best ways I can think of for replicating such a warning in miette are:

<details><summary>Rust source</summary>

```rust
use miette::{Diagnostic, GraphicalReportHandler, NamedSource, ReportHandler, SourceSpan};
use std::fmt::{self, Display};
use thiserror::Error;

#[derive(Debug, Error, Diagnostic)]
#[error("unreachable `pub` item")]
#[diagnostic(
    severity(warning),
    help("or consider exporting it for use by other crates")
)]
struct UnreachablePub {
    #[source_code]
    code: NamedSource,
    #[label]
    item: SourceSpan,
    #[label("consider restricting its visibility: `pub(crate)`")]
    pub_: SourceSpan,
    #[related]
    lint: Option<LintDef>,
}

#[derive(Debug, Error, Diagnostic)]
#[error("the lint level is defined here")]
#[diagnostic(severity(advice))]
struct LintDef {
    #[label]
    span: SourceSpan,
}

fn main() {
    let src = "\
#![warn(unreachable_pub)]
pub use crate::{ExpactSpanTrace as _, InstrumentError as _, InstrumentResult as _};
";

    let diagnostic = UnreachablePub {
        code: NamedSource::new("src/lib.rs", src),
        item: (42..62).into(),
        pub_: (26..29).into(),
        lint: Some(LintDef {
            span: (8..23).into(),
        }),
    };

    println!("{}", GraphicalDiagnisticDisplay(&diagnostic));
}

struct GraphicalDiagnisticDisplay<'a>(&'a dyn Diagnostic);

impl Display for GraphicalDiagnisticDisplay<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        GraphicalReportHandler::new().debug(self.0, f)
    }
}
```
</details>

```
  ⚠ unreachable `pub` item
   ╭─[src/lib.rs:1:1]
 1 │ #![warn(unreachable_pub)]
 2 │ pub use crate::{ExpactSpanTrace as _, InstrumentError as _, InstrumentResult as _};
   · ─┬─             ────────────────────
   ·  ╰── consider restricting its visibility: `pub(crate)`
   ╰────
  help: or consider exporting it for use by other crates

Error:
  ☞ the lint level is defined here
   ╭─[src/lib.rs:1:1]
 1 │ #![warn(unreachable_pub)]
   ·         ───────────────
 2 │ pub use crate::{ExpactSpanTrace as _, InstrumentError as _, InstrumentResult as _};
   ╰────
```

or maybe

<details><summary>Rust source</summary>

```rust
use miette::{Diagnostic, GraphicalReportHandler, NamedSource, ReportHandler, SourceSpan};
use std::fmt::{self, Display};
use thiserror::Error;

#[derive(Debug, Error, Diagnostic)]
#[error("unreachable `pub` item")]
#[diagnostic(
    severity(warning),
    help("or consider exporting it for use by other crates")
)]
struct UnreachablePub {
    #[source_code]
    code: NamedSource,
    #[label]
    item: SourceSpan,
    #[label("consider restricting its visibility: `pub(crate)`")]
    pub_: SourceSpan,
    #[label("the lint level is defined here")]
    lint: SourceSpan,
}

fn main() {
    let src = "\
#![warn(unreachable_pub)]
pub use crate::{ExpactSpanTrace as _, InstrumentError as _, InstrumentResult as _};
";

    let diagnostic = UnreachablePub {
        code: NamedSource::new("src/lib.rs", src),
        item: (42..62).into(),
        pub_: (26..29).into(),
        lint: (8..23).into(),
    };

    println!("{}", GraphicalDiagnisticDisplay(&diagnostic));
}

struct GraphicalDiagnisticDisplay<'a>(&'a dyn Diagnostic);

impl Display for GraphicalDiagnisticDisplay<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        GraphicalReportHandler::new().debug(self.0, f)
    }
}
```
</details>

```
  ⚠ unreachable `pub` item
   ╭─[src/lib.rs:1:1]
 1 │ #![warn(unreachable_pub)]
   ·         ───────┬───────
   ·                ╰── the lint level is defined here
 2 │ pub use crate::{ExpactSpanTrace as _, InstrumentError as _, InstrumentResult as _};
   · ─┬─             ────────────────────
   ·  ╰── consider restricting its visibility: `pub(crate)`
   ╰────
  help: or consider exporting it for use by other crates
```

This issue honestly might just be "The flaws in `#[related]`'s design."

**Comment by CAD97** [comment]
There's two _main_ uses for something `#[related]`-adjacent that I can think of:

- One logical diagnostic which has multiple parts, e.g.

  ```
    ⚠ unreachable `pub` item
     ╭─[src/lib.rs:2:16]
   2 │ pub use crate::{ExpactSpanTrace as _, InstrumentError as _, InstrumentResult as _};
     · ───             ────────────────────
     ╰────
    ☞ the lint level is defined here
     ╭─[src/lib.rs:1:8]
   1 │ #![warn(unreachable_pub)]
     ·    ──── ───────────────
     ╰────
    help: consider restricting its visibility: `pub(crate)`
    help: or consider exporting it for use by other crates
  ```

- A container of multiple logically distinct diagnostics being reported together as the result of one process, e.g.

  ```
  Error:
    × compilation failed
  
  Error: D0000 (link)
    × I won't allow this
     ╭─[1:11]
   1 │ this is an error
     ·            ─────
     ╰────
  
  Warn: D4000 (link)
    ⚠ I don't like this
     ╭─[2:10]
   2 │ this is a warning
     ·           ───────
     ╰────
  
  Info: D8000 (link)
    ☞ This could be better
     ╭─[3:11]
   3 │ this is an advice
     ·            ──────
     ╰────
  ```

The former feels like something miette should address[^1]. The latter feels like something nice-to-have for initially getting up and running, but will usually need to be replaced by something more custom{^2].

[^1]: Proposal: `#[also]`: takes a single `impl Diagnostic` (unlike `#[related]` which takes an iterable) to display directly after the main diagnostic, as "part of" the primary diagnostic. As such, the graphical handler does not print the `also` diagnostic's code and puts no blank lines between the primary part and the also part.

[^2]: Or in other words: `#[related]` is fine and can stay. The only changes I personally would suggest making are[^3] <ul><li>Print an empty line between diagnostics, to improve grouping; and</li><li></li>Change the `Error:` introducer based on the severity of the related diagnostic.</ul>

[^3]: Ugh doing _anything_ not a single paragraph in footnotes requires dropping down into raw HTML. This is footnote abuser abuse /sarcasm!

**Comment by zkat** [comment]
hey, sorry for the delayed response.

I agree with your assessment, and I'm currently thinking of ways to introduce something for `#1` there without complicating the protocol too much, and I think I have some idea how to do that. I'll give it a shot and get back to you.

As far as the `#[related]` issue goes, I think it's prudent to take a patch such that related errors print out the appropriate prefix. Would you mind putting that together?

**Comment by zkat** [comment]
We can keep talking about what to do about Related or similar elsewhere, but I've fixed the immediate issue here and released it with miette@4.6.0 :)
